packages/salesforce-adapter/src/filters/field_permissions_enum.ts

```diff
@@ -0,0 +1,282 @@
+  if (isTypeWithFieldPermissions(objectType)) {
+    objectType.fields.fieldPermissions.refType = getMapOfMapOfFieldPermissionEnum()
+  }
+}
+
+const fieldPermissionsFieldToOriginalType = (objectType: ObjectType): void => {
+  if (isTypeWithFieldPermissions(objectType)) {
+    objectType.fields.fieldPermissions.refType = getMapOfMapOfProfileFieldLevelSecurity()
+  }
+}
+
+let shouldRunDeployFilters: boolean
+
+// The decision if to run the deploy filters is based on the fieldPermissions type
+// which indicates if it this filter ran onFetch or not
+const shouldRunDeployFiltersAccordingToInstanceType = async (
+  instanceType: ObjectType,
+): Promise<boolean> =>
+  (isTypeWithFieldPermissions(instanceType)
+    && (await instanceType.fields.fieldPermissions.getType()).elemID
+      .isEqual(getMapOfMapOfFieldPermissionEnum().elemID))
+
+
+const filter: LocalFilterCreator = ({ config }) => ({
+  onFetch: async elements => {
+    if (config.enumFieldPermissions === false) {
+      return
+    }
+    log.info('Running enumFieldPermissions onFetch')
+    await awu(metadataTypesWithFieldPermissions).forEach(async metadataType => {
+      const type = await awu(elements).find(
+        async element => isObjectType(element) && await apiName(element) === metadataType,
+      ) as ObjectType
+      if (type !== undefined) {
+        fieldPermissionFieldToEnum(type)
+      }
+    })
+    const instancesWithFieldPermissions = await awu(elements)
+      .filter(isInstanceOfType(...metadataTypesWithFieldPermissions))
+      .toArray() as InstanceElement[]
+    instancesWithFieldPermissions.forEach(fieldPermissionValuesToEnum)
+    elements.push(enumFieldPermissions)
+  },
+  preDeploy: async changes => {
+    const instanceChangesWithFieldPermissions = await awu(changes)
+      .filter(isAdditionOrModificationChange)
+      .filter(isInstanceOfTypeChange(...metadataTypesWithFieldPermissions))
+      .toArray() as (AdditionChange<InstanceElement> | ModificationChange<InstanceElement>)[]
+    if (instanceChangesWithFieldPermissions.length === 0) {
+      return
+    }
+    const instanceType = await getChangeData(instanceChangesWithFieldPermissions[0]).getType()
+    shouldRunDeployFilters = await shouldRunDeployFiltersAccordingToInstanceType(instanceType)
+    if (!shouldRunDeployFilters) {
+      return
+    }
+    log.info('Running enumFieldPermissions preDeploy')
+    instanceChangesWithFieldPermissions.forEach(instanceChange => (
+      applyFunctionToChangeData(
+        instanceChange,
+        fieldPermissionValuesToObject,
+      )
+    ))
+    const metadataTypeToInstanceChanges = await awu(instanceChangesWithFieldPermissions)
+      .groupBy(async instanceChange =>
+        apiName(await getChangeData(instanceChange).getType()))
+    const instanceTypes = await awu(Object.values(metadataTypeToInstanceChanges))
+      .map(async instanceChanges => getChangeData(instanceChanges[0]).getType())
+      .toArray()
```

Comment:
less code to do the same thing?
```suggestion
    const instanceTypes = Object.values(
      await awu(instanceChangesWithFieldPermissions.map(getChangeData))
        .map(inst => inst.getType())
        .keyBy(type => apiName(type))
    )
```